crates/wasmtime/src/module.rs

```diff
@@ -329,6 +317,27 @@ impl Module {
         Self::from_parts(engine, modules, main_module, Arc::new(types), &[])
     }
 
+    /// Deserializes an in-memory compiled module previously created with
+    /// [`Module::serialize`] or [`Engine::precompile_module`].
+    ///
+    /// This function will deserialize the binary blobs emitted by
+    /// [`Module::serialize`] and [`Engine::precompile_module`] back into an
+    /// in-memory [`Module`] that's ready to be instantiated. Note that this is
+    /// only compatible with bytes that were produced by the same version of the
+    /// wasmtime crate itself.
+    ///
+    /// It's important to not that this is somewhat `unsafe` but is not marked
+    /// as `unsafe`. It is technically possible to feed in arbitrary bytes here
+    /// and have custom jit code which enables arbitrary code execution. The
+    /// `bytes` structure is validated to have the right "structure" of bytes
+    /// but the contents are not validated by `wasmtime`. It is safe to pass in
+    /// any wasmtime-compiled blob here (from future and past versions), but
+    /// this should not be exposed to arbitrary user-defined input.
```

Comment:
This is not really convincing me that this should _not_ be marked `unsafe`. It seems like this is describing the contract one must uphold for it to be safe, that it must have be some wasmtime-compiled blob, and the whole point of `unsafe` is encapsulating such external-to-the-type-system safety invariants.

Reply:
I was originally going to mark this `unsafe`, but the impact on other language bindings gave me hesitation. Languages like Python expect everything to be safe, so this sort of documentation would inevitably end up there  as well. Or otherwise this would need to show up everywhere anyway (basically the `unsafe` in Rust doesn't help any users anywhere else).

In the end I thought of this along the lines of `File::open` and `/dev/self/mem` (or whatever it's called). Technically it's unsafe but practically it's never used that way.

Reply:
I could go either way on this issue; it's unsafe for arbitrary input but safe for all desired use cases (i.e. rehydrating precompiled/serialized output).  Either way (marked unsafe or not) we'll want the big, bold caveat of "it's the caller's responsibility to ensure integrity of the serialized data before calling this method".

Reply:
I guess I would prefer to mark it as `unsafe` and also propagate that warning about "only use this with files you created via wasmtime directly" out to the language embeddings too.

> In the end I thought of this along the lines of `File::open` and `/dev/self/mem` (or whatever it's called). Technically it's unsafe but practically it's never used that way.

I see the argument, but I think this is a different category of thing: with `/dev/self/mem` we have a :sparkles: magical :sparkles: file, but with `Module::deserialize` regular old files can trip us up if they have the right shape.

Reply:
Hm ok I'll go with `unsafe` and we can see how it goes.